azure/components/userpool/state.ftl

```diff
@@ -0,0 +1,86 @@
+[#ftl]
+
+[#macro azure_userpool_arm_state occurrence parent={}]
+    
+    [#local core = occurrence.Core]
+    [#local solution = occurrence.Configuration.Solution]
+
+    [#assign componentState = 
+        {
+            "Resources" : {},
+            "Attributes" : {
+                "API_AUTHORIZATION_HEADER" : solution.AuthorizationHeader
+            },
+            "Roles" : {}
+        }
+    ]
+
+[/#macro]
+
+[#macro azure_userpoolclient_arm_state occurrence parent={}]
+    
+    [#local core = occurrence.Core]
+    [#local solution = occurrence.Configuration.Solution]
+
+    [#local id = formatResourceId(AZURE_APPLICATION_REGISTRATION_RESOURCE_TYPE, core.Name)]
+    [#local name = formatName(core.Name)]
+
+    [#local parentAttributes = parent.State.Attributes]
+
+    [#assign componentState=
+        {
+            "Resources" : {
+                "client" : {
+                    "Id" : id,
+                    "Name" : name,
+                    "Type" : AZURE_APPLICATION_REGISTRATION_RESOURCE_TYPE,
+                    "Reference" : getReference(id, name)
+                }
+            },
+            "Attributes" : parentAttributes + {
+                "CLIENT" : getExistingReference(id)
+            },
+            "Roles" : {
+                "Inbound" : {},
+                "Outbound" : {}
+            }
+        }
+    ]
+
+[/#macro]
+
+[#macro azure_userpoolauthprovider_arm_state occurrence parent={}]
+
+    [#local core = occurrence.Core]
+    [#local name = core.Name]
+
+    [#local validProviders = [
+        "facebook", 
+        "google", 
+        "microsoft", 
+        "twitter", 
+        "aad", 
+        "aadb2c"]
```

Comment:
Might be better to set the provider using the Engine attribute rather than the name. Using an attribute allows it do be documented in our component reference rather than being treated as an exception.

Reply:
Updated. Implemented Azure-specific Engine attributes and validation.